modules/calcite/src/main/java/org/apache/ignite/internal/processors/query/calcite/exec/rel/AggregateSortNode.java

```diff
@@ -0,0 +1,299 @@
+import org.apache.ignite.internal.processors.query.calcite.exec.RowHandler.RowFactory;
+import org.apache.ignite.internal.processors.query.calcite.exec.exp.agg.Accumulator;
+import org.apache.ignite.internal.processors.query.calcite.exec.exp.agg.AccumulatorWrapper;
+import org.apache.ignite.internal.processors.query.calcite.exec.exp.agg.AggregateType;
+import org.apache.ignite.internal.processors.query.calcite.util.Commons;
+import org.apache.ignite.internal.util.typedef.F;
+
+/**
+ *
+ */
+public class AggregateSortNode<Row> extends AbstractNode<Row> implements SingleNode<Row>, Downstream<Row> {
+    /** */
+    private final AggregateType type;
+
+    /** */
+    private final Supplier<List<AccumulatorWrapper<Row>>> accFactory;
+
+    /** */
+    private final RowFactory<Row> rowFactory;
+
+    /** */
+    private final ImmutableBitSet grpSet;
+
+    /** */
+    private final Comparator<Row> comp;
+
+    /** */
+    private Row prevRow;
+
+    /** */
+    private Group grp;
+
+    /** */
+    private int requested;
+
+    /** */
+    private int waiting;
+
+    /** */
+    private int cmpRes;
+
+    /**
+     * @param ctx Execution context.
+     */
+    public AggregateSortNode(
+        ExecutionContext<Row> ctx,
+        RelDataType rowType,
+        AggregateType type,
+        ImmutableBitSet grpSet,
+        Supplier<List<AccumulatorWrapper<Row>>> accFactory,
+        RowFactory<Row> rowFactory,
+        Comparator<Row> comp
+    ) {
+        super(ctx, rowType);
+
+        this.type = type;
+        this.accFactory = accFactory;
+        this.rowFactory = rowFactory;
+        this.grpSet = grpSet;
+        this.comp = comp;
+    }
+
+    /** {@inheritDoc} */
+    @Override public void request(int rowsCnt) {
+        assert !F.isEmpty(sources()) && sources().size() == 1;
+        assert rowsCnt > 0 && requested == 0;
+        assert waiting <= 0;
+
+        try {
+            checkState();
+
+            requested = rowsCnt;
+
+            if (waiting == 0)
+                source().request(waiting = IN_BUFFER_SIZE);
+        }
+        catch (Exception e) {
+            onError(e);
+        }
+    }
+
+    /** {@inheritDoc} */
+    @Override public void push(Row row) {
+        assert downstream() != null;
+        assert waiting > 0;
+
+        try {
+            checkState();
+
+            waiting--;
+
+            if (grp != null) {
+                int cmp = comp.compare(row, prevRow);
+
+                if (cmp == 0)
+                    grp.add(row);
+                else {
+                    if (cmpRes == 0)
+                        cmpRes = cmp;
+                    else
+                        assert cmp == cmpRes : "Input not sorted";
+
+                    doPush();
+
+                    grp = newGroup(row);
+                }
+            }
+            else
+                grp = newGroup(row);
+
+            prevRow = row;
+
+            if (waiting == 0)
+                source().request(waiting = IN_BUFFER_SIZE);
```

Comment:
it's better to request new batch in another task to allow other queries do progress